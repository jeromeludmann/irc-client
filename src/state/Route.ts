import { Action } from "redux";

export interface Route {
  server: string;
  channel: string;
}

export interface Routable extends Action<string> {
  route: Route;
}
