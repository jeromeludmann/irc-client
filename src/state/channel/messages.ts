import { SEND_INPUT_VALUE, SendInputValueAction } from "@app/actions/ui/input";
import { JoinAction, MESSAGE_JOIN } from "@app/actions/messages/join";
import { NoticeAction, MESSAGE_NOTICE } from "@app/actions/messages/notice";
import { PingAction, MESSAGE_PING } from "@app/actions/messages/ping";
import { PrivmsgAction, MESSAGE_PRIVMSG } from "@app/actions/messages/privmsg";

export type MessageListState = string[];

export type MessageListAction =
  | SendInputValueAction
  | JoinAction
  | NoticeAction
  | PingAction
  | PrivmsgAction;

export const messagesInitialState: MessageListState = [];

export default function reduceMessages(
  messages = messagesInitialState,
  action: MessageListAction,
): MessageListState {
  switch (action.type) {
    case SEND_INPUT_VALUE: {
      return [...messages, ">>> " + action.payload.value];
    }
    case MESSAGE_JOIN: {
      const { user, channel } = action.payload;
      const msg = `${user.nick} has joined ${channel}`;
      return [...messages, msg];
    }
    case MESSAGE_PING: {
      const { server } = action.payload;
      return [...messages, `Received ping from ${server}`];
    }
    case MESSAGE_PRIVMSG: {
      const { user, text } = action.payload;
      const msg = `${user.nick}: ${text}`;
      return [...messages, msg];
    }
    case MESSAGE_NOTICE: {
      return [...messages, action.payload.text];
    }
    default:
      return messages;
  }
}
