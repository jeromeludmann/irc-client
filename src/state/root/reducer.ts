import { RoutedAction, isChannel, isPrivate } from '@app/utils/Route'
import {
  RouteState,
  routeInitialState,
  reduceRoute,
} from '@app/state/route/reducer'
import {
  ServerState,
  reduceServer,
  serverInitialState,
} from '@app/state/server/reducer'
import { CLOSE_WINDOW, ADD_NEW_SERVER } from '@app/actions/ui'

type RootPartialState = Readonly<{
  servers: Readonly<{ [key: string]: ServerState }>
}>

export type RootState = {
  route: RouteState
} & RootPartialState

type RootReducer<S = RootState> = (
  root: S,
  action: RoutedAction,
  extraStates: { root: RootState },
) => S

export const rootInitialState = {
  servers: { serverKey: serverInitialState },
  route: routeInitialState,
}

const caseReducers: { [action: string]: RootReducer<RootPartialState> } = {
  [CLOSE_WINDOW]: (root, action, _) => {
    const thereIsOnlyOneServer = Object.keys(root.servers).length <= 1
    const isChannelOrPrivate =
      isChannel(action.route.bufferKey) || isPrivate(action.route.bufferKey)

    if (thereIsOnlyOneServer || isChannelOrPrivate) {
      return root
    }

    const servers = { ...root.servers }
    delete servers[action.route.serverKey]
    return { servers }
  },
}

const routeActionToServers = (
  servers: { [key: string]: ServerState },
  action: RoutedAction,
  extraStates: { route: RouteState },
) =>
  action.route.serverKey in servers || action.type === ADD_NEW_SERVER
    ? {
        ...servers,
        [action.route.serverKey]: reduceServer(
          servers[action.route.serverKey],
          action,
          extraStates,
        ),
      }
    : servers

export function reduceRoot(
  root: RootState | undefined = rootInitialState,
  action: RoutedAction,
): RootState {
  // prevent other actions to pass inside app reducers
  if (action.route === undefined) {
    return root
  }

  const intermediateState = {
    ...root,
    route: reduceRoute(root.route, action, { root }),
    servers: routeActionToServers(root.servers, action, {
      route: root.route,
    }),
  }

  return {
    ...intermediateState,
    ...(action.type in caseReducers
      ? caseReducers[action.type](intermediateState, action, {
          root: intermediateState,
        })
      : {}),
  }
}
