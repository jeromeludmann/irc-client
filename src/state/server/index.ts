import { Action } from "redux";
import reduceModes, {
  ModesState,
  modesInitialState,
} from "@app/state/server/modes";
import reduceChannels, {
  ChannelRouterState,
  ChannelRouterAction,
  channelsInitialState,
} from "@app/state/server/channel-router";
import { Route } from "@app/state/Route";
import { ActiveRouteState } from "@app/state/active";

export type ServerState = {
  readonly modes: ModesState;
  readonly channels: ChannelRouterState;
};

export type ServerAction = Action;

interface ExtraParams {
  route: Route;
  active: ActiveRouteState;
}
export const serverInitialState: ServerState = {
  channels: channelsInitialState,
  modes: modesInitialState,
};

export default function reduceServer(
  server = serverInitialState,
  action: ServerAction,
  { route, active }: ExtraParams,
): ServerState {
  return {
    modes: reduceModes(server.modes),
    channels: reduceChannels(server.channels, action as ChannelRouterAction, {
      route,
      active,
    }),
  };
}
