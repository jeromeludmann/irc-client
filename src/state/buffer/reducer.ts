import {
  MessagesState,
  messagesInitialState,
  reduceMessages,
} from '@app/state/messages/reducer'
import {
  InputState,
  inputInitialState,
  reduceInput,
} from '@app/state/input/reducer'
import { RoutedAction } from '@app/utils/Route'
import { RouteState } from '@app/state/route/reducer'
import { RECEIVE_JOIN, ReceiveJoinAction } from '@app/actions/messages/incoming'
import { SWITCH_WINDOW } from '@app/actions/ui'
import { ServerState } from '@app/state/server/reducer'
import { AnyAction } from 'redux'

type BufferPartialState = Readonly<{
  activity: boolean
}>

export type BufferState = Readonly<{
  input: InputState
  messages: MessagesState
}> &
  BufferPartialState

type BufferReducer<S = BufferState> = (
  buffer: S,
  action: RoutedAction,
  extraStates: {
    route: RouteState
    server: ServerState
  },
) => S

export const bufferInitialState = {
  activity: false,
  input: inputInitialState,
  messages: messagesInitialState,
}

const itIsMe = (action: AnyAction, extraStates: { server: ServerState }) =>
  action.payload.user.nick === extraStates.server.user.nick

const caseReducers: { [action: string]: BufferReducer<BufferPartialState> } = {
  [RECEIVE_JOIN]: (_, action: ReceiveJoinAction, extraStates) => ({
    activity: !itIsMe(action, extraStates),
  }),

  [SWITCH_WINDOW]: () => ({ activity: false }),
}

export const reduceBuffer: BufferReducer = (
  buffer,
  action,
  extraStates,
): BufferState => ({
  ...(action.type in caseReducers
    ? caseReducers[action.type](buffer, action, extraStates)
    : buffer),
  input: reduceInput(buffer.input, action),
  messages: reduceMessages(buffer.messages, action, extraStates),
})
