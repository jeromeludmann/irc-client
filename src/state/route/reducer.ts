import { Action, AnyAction } from 'redux'
import { Route, BufferKey, isStatus } from '@app/utils/Route'
import {
  SWITCH_WINDOW,
  SwitchWindowAction,
  CLOSE_WINDOW,
} from '@app/actions/ui'
import {
  ReceiveJoinAction,
  RECEIVE_JOIN,
  ReceivePartAction,
  RECEIVE_PART,
} from '@app/actions/messages/incoming'
import { RootState } from '@app/state/root/reducer'

export type RouteState = Readonly<Route>

export const routeInitialState: RouteState = {
  serverKey: 'serverKey',
  bufferKey: BufferKey.STATUS,
}

type RouteReducer = (
  state: RouteState,
  action: Action,
  extraStates: { root: RootState },
) => RouteState

const itIsMe = (action: AnyAction, extraStates: { root: RootState }) =>
  action.payload.user.nick ===
  extraStates.root.servers[action.route.serverKey].user.nick

const caseReducers: { [action: string]: RouteReducer } = {
  [CLOSE_WINDOW]: (route, _, extraStates) => {
    if (isStatus(route.bufferKey)) {
      const keys = Object.keys(extraStates.root.servers)
      return keys.length > 1
        ? {
            serverKey: keys.filter(key => key !== route.serverKey)[0],
            bufferKey: BufferKey.STATUS,
          }
        : route
    }

    return { ...route, bufferKey: BufferKey.STATUS }
  },

  [RECEIVE_JOIN]: (route, action: ReceiveJoinAction, extraStates) =>
    itIsMe(action, extraStates) ? action.route : route,

  [RECEIVE_PART]: (route, action: ReceivePartAction, extraStates) =>
    itIsMe(action, extraStates)
      ? { ...route, bufferKey: BufferKey.STATUS }
      : route,

  [SWITCH_WINDOW]: (route, action: SwitchWindowAction, extraStates) => {
    const { serverKey, bufferKey } = action.route

    const routeFound =
      serverKey in extraStates.root.servers &&
      bufferKey in extraStates.root.servers[serverKey].buffers

    if (!routeFound) {
      console.log(`Route "${JSON.stringify(action.route)}" not found`)
      return route
    }

    return action.route
  },
}

export const reduceRoute: RouteReducer = (
  route = routeInitialState,
  action,
  extraStates,
) =>
  action.type in caseReducers
    ? caseReducers[action.type](route, action, extraStates)
    : route
