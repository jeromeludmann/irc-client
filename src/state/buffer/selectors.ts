import { createSelector } from 'reselect'
import { selectBuffers } from '@app/state/server/selectors'
import { selectRoute } from '@app/state/route/selectors'

export const selectBuffer = createSelector(
  selectBuffers,
  selectRoute,
  (buffers, route) => buffers[route.bufferKey],
)

export const selectActivity = createSelector(
  selectBuffer,
  buffer => buffer.activity,
)
