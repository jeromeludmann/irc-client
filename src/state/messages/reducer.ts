import { Action } from 'redux'
import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RawMessagesAction,
  CONNECTION_CLOSED,
  RAW_MESSAGES_RECEIVED,
  CONNECT_TO_SERVER,
  ConnectToServerAction,
} from '@app/actions/socket'
import {
  ReceiveErrorAction,
  ReceiveJoinAction,
  ReceivePartAction,
  ReceivePrivmsgAction,
  ReceiveNoticeFromServerAction,
  ReceiveNoticeFromChannelAction,
  ReceiveNoticeFromUserAction,
  RECEIVE_ERROR,
  RECEIVE_JOIN,
  RECEIVE_NOTICE_FROM_SERVER,
  RECEIVE_NOTICE_FROM_CHANNEL,
  RECEIVE_NOTICE_FROM_USER,
  RECEIVE_PART,
  RECEIVE_PRIVMSG,
  RECEIVE_PING_FROM_SERVER,
} from '@app/actions/messages/incoming'
import {
  SEND_PONG_TO_SERVER,
  SEND_PRIVMSG,
  SendPrivmsgAction,
} from '@app/actions/messages/outgoing'
import {
  PRINT_HELP_BY_DEFAULT,
  PrintHelpByDefaultAction,
  PRINT_HELP_ABOUT_COMMAND,
  PrintHelpAboutCommandAction,
} from '@app/actions/commands'
import { RouteState } from '@app/state/route/reducer'
import { ServerState } from '@app/state/server/reducer'
import { UPDATE_SERVER_LAG, UpdateServerLagAction } from '@app/actions/ui'

// TODO replace string[] by MessageState[]
export type MessagesState = string[]

type MessagesReducer = (
  messages: MessagesState,
  action: Action,
  extraStates: { route: RouteState; server: ServerState },
) => MessagesState

export const messagesInitialState: MessagesState = []

const caseReducers: { [action: string]: MessagesReducer } = {
  [CONNECT_TO_SERVER]: (messages, action: ConnectToServerAction) => [
    ...messages,
    `Connecting to ${action.payload.host}â€¦`,
  ],

  [CONNECTION_CLOSED]: messages => [
    ...messages,
    'Disconnected from remote host.',
  ],

  [CONNECTION_FAILED]: (
    messages,
    { payload: { message } }: ConnectionFailedAction,
  ) => [...messages, message],

  [PRINT_HELP_BY_DEFAULT]: (
    messages,
    { payload: { commands } }: PrintHelpByDefaultAction,
  ) => [
    ...messages,
    ...Object.keys(commands).map(
      commandName => `${commandName} - ${commands[commandName].description}`,
    ),
    'Type /HELP <command> for more details.',
  ],

  [PRINT_HELP_ABOUT_COMMAND]: (
    messages,
    { payload: { command } }: PrintHelpAboutCommandAction,
  ) => [
    ...messages,
    `Usage: /${command.name} ${command.syntax} - ${command.description}`,
  ],

  [RAW_MESSAGES_RECEIVED]: (messages, action: RawMessagesAction) => [
    ...messages,
    ...action.payload.messages,
  ],

  [RECEIVE_ERROR]: (messages, { payload: { message } }: ReceiveErrorAction) => [
    ...messages,
    message,
  ],

  [RECEIVE_JOIN]: (
    messages,
    { payload: { user, channel } }: ReceiveJoinAction,
  ) => [...messages, `${user.nick} has joined ${channel}`],

  [RECEIVE_NOTICE_FROM_SERVER]: (
    messages,
    action: ReceiveNoticeFromServerAction,
  ) => [...messages, action.payload.text],

  [RECEIVE_NOTICE_FROM_CHANNEL]: (
    messages,
    {
      payload: { user, text },
      route: { bufferKey },
    }: ReceiveNoticeFromChannelAction,
  ) => [...messages, `-${user.nick}/${bufferKey}- ${text}`],

  [RECEIVE_NOTICE_FROM_USER]: (
    messages,
    { payload: { user, text } }: ReceiveNoticeFromUserAction,
  ) => [...messages, `-${user.nick}- ${text}`],

  [RECEIVE_PART]: (
    messages,
    { payload: { user, channel, message } }: ReceivePartAction,
  ) => {
    const baseMsg = `${user.nick} has left ${channel}`
    return [...messages, message ? `${baseMsg} (${message})` : baseMsg]
  },

  [RECEIVE_PING_FROM_SERVER]: messages => [...messages, 'Ping?'],

  [RECEIVE_PRIVMSG]: (
    messages,
    { payload: { user, text } }: ReceivePrivmsgAction,
  ) => [...messages, `${user.nick}: ${text}`],

  [SEND_PONG_TO_SERVER]: messages => [...messages, 'Pong!'],

  [SEND_PRIVMSG]: (messages, action: SendPrivmsgAction, { server }) => [
    ...messages,
    `${server.user.nick}: ${action.payload.text}`,
  ],

  [UPDATE_SERVER_LAG]: (
    messages,
    { payload: { lag } }: UpdateServerLagAction,
    {},
  ) => [...messages, `Server lag: ${lag} ms`],
}

export const reduceMessages: MessagesReducer = (
  messagesState = messagesInitialState,
  action,
  extraStates,
) =>
  action.type in caseReducers
    ? caseReducers[action.type](messagesState, action, extraStates)
    : messagesState
