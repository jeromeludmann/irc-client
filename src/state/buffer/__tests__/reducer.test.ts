import { reduceBuffer, bufferInitialState } from '@app/state/buffer/reducer'
import { messageReceivers } from '@app/actions/messages/incoming'
import { User } from '@app/utils/Message'
import { serverInitialState } from '@app/state/server/reducer'
import { switchWindow } from '@app/actions/ui'
import { Route } from '@app/utils/Route'

describe('reduce buffer state', () => {
  const someone: User = { nick: 'someone', user: 'user', host: 'host' }
  const route: Route = { serverKey: 'serverKey', bufferKey: '#channel' }
  const extraStates = { server: serverInitialState, route }

  it('should handle RECEIVE_JOIN', () => {
    expect(
      reduceBuffer(
        undefined!,
        messageReceivers.JOIN('serverKey', someone, ['#channel']),
        extraStates,
      ),
    ).toMatchSnapshot()
  })

  it('should handle SWITCH_WINDOW', () => {
    expect(
      reduceBuffer(bufferInitialState, switchWindow(route), extraStates),
    ).toMatchSnapshot()
  })
})
