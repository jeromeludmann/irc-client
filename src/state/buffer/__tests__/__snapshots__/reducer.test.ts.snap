// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduce buffer state should handle RECEIVE_JOIN 1`] = `
Object {
  "activity": true,
  "input": Object {
    "dirtyValue": "",
    "history": Object {
      "index": 0,
      "values": Array [],
    },
    "value": "",
  },
  "messages": Array [
    "someone has joined #channel",
  ],
}
`;

exports[`reduce buffer state should handle SWITCH_WINDOW 1`] = `
Object {
  "activity": false,
  "input": Object {
    "dirtyValue": "",
    "history": Object {
      "index": 0,
      "values": Array [],
    },
    "value": "",
  },
  "messages": Array [],
}
`;
